---
title: "PopGenome"
author: "PP"
date: "2024-06-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(PopGenome)
library(ggplot2)
library(tidyverse)
library(rtracklayer)
library(vcfR)
```

```{r}
filter = "YES"
if (filter == "YES"){
  vcf <- read.vcfR("data/vcfs/Singhal_bTG1_4_Z_plusLTF_filtered_snps.vcf.gz")
  gff <- import.gff("data/gffs/NC_044241.2.gff.gz")
  coding_regions <- gff[gff$type == "CDS"]
  
  
  filter_vcf_func <- function(snps, start, width){
    end <- start + width-1
    snps <- as.numeric(snps)
    rm_positions <- snps[which(snps > start & snps < end)]
    return(rm_positions)
  }
  
  
  snps <- vcf@fix[,2] %>% as.numeric()
  
  rm_snps <- coding_regions@ranges %>% as.data.frame() %>% 
    apply(., 1, function(x)(filter_vcf_func(snps, x[1], width = x[3]))) %>% unlist() %>% 
    unique()
  
  rms_indx <- which(snps %in% rm_snps)
  vcf@gt <- vcf@gt[-rms_indx,]
  vcf@fix <- vcf@fix[-rms_indx,]
  
  write.vcf(vcf, "data/vcfs/Z_nocoding.vcf.gz")
}
```


```{bash engine.opts='-l'}
# This is a bash code chunk
conda activate vcftools
vcftools --gzvcf data/vcfs/Z_nocoding.vcf.gz \
  --recode --out data/vcfs/Z_nocoding_filt \
  --minDP 5 --minGQ 20 \
  --max-missing 0.5
  
bgzip -f -c data/vcfs/Z_nocoding_filt.recode.vcf > data/vcfs/Z_nocoding_filt.vcf.gz
rm data/vcfs/Z_nocoding_filt.recode.vcf
tabix -p vcf data/vcfs/Z_nocoding_filt.vcf.gz

vcftools --gzvcf data/vcfs/Singhal_bTG1_4_Z_plusLTF_filtered_snps.vcf.gz  \
  --recode --out data/vcfs/Z_filt \
  --minDP 5 --minGQ 20 \
  --max-missing 0.5
  
bgzip -f -c data/vcfs/Z_filt.recode.vcf > data/vcfs/Z_filt.vcf.gz
rm data/vcfs/Z_filt.recode.vcf
tabix -p vcf data/vcfs/Z_filt.vcf.gz


conda deactivate
```




```{r}
### SAMPLE INFO ###
sample_info_ENI <- read.table("data/filereport_read_run_PRJEB10586_tsv.txt", sep = 
                                '\t', header = T) %>% 
  select(run_accession, sample_accession, scientific_name)

sample_info_jake <- read.csv("data/jake_sample_info.csv")

sample_info <- sample_info_ENI %>% 
  merge(sample_info_jake, by.x = 'sample_accession', by.y = 'Sample.Accession') %>% 
  filter(Karyotype %in% c("A", "AA", "AB", "BB", "B"))


MAs <- sample_info[sample_info$Karyotype %in% c("AA") & sample_info$Sex == "M",]$run_accession
MAs2 <- paste0(MAs, ".2")
FAs <- sample_info[sample_info$Karyotype %in% c("A") & sample_info$Sex == "F",]$run_accession

MBs <- sample_info[sample_info$Karyotype %in% c("BB", "B") & sample_info$Sex == "M",]$run_accession
MBs2 <- paste0(MBs, ".2")
FBs <- sample_info[sample_info$Karyotype %in% c("BB", "B") & sample_info$Sex == "F",]$run_accession

A <- c(MAs, MAs2, FAs)
B <- c(MBs, MBs2, FBs)
A_ss <- sample(A, length(B), replace = F)
```


```{r}
#### pseudohaploiding the vcfs ####
vcfs <- c("data/vcfs/Z_nocoding_filt.vcf.gz", 
          "data/vcfs/Z_filt.vcf.gz", 
          "data/vcfs/Singhal_bTG1_4_Z_plusLTF_filtered_snps.vcf.gz")

for (vcf in vcfs){
  vcf_whole <- read.vcfR(vcf)
  
  ### GETTING SEX SAMPLES ###
  Females <- intersect(sample_info$run_accession[sample_info$Sex == "F"], colnames(vcf_whole@gt))
  Males <- c(intersect(sample_info$run_accession[sample_info$Sex == "M"], colnames(vcf_whole@gt)), "LTF")
  
  ### GETTING GT FUNCTION ###
  get_gt <- function(samp, gt){
    gs <- str_split(gt[,samp], ":", simplify = T)[,1]
    df <- data.frame(genotype = gs)
    colnames(df) <- samp
    return(df)
  }
  
  ### MALE GENOTYPES ###
  just_gt_Males <- lapply(Males, get_gt, gt = vcf_whole@gt) %>% bind_cols() 
  
  ### FEMALE GENOTYPES ###
  just_gt_Females <- lapply(Females, get_gt, gt = vcf_whole@gt) %>% bind_cols()
  
  ### FUNCTION TO CONVERT DIPLOID MALES TO PSEUDOHAPLOID ###
  hap_the_dip <- function(samp, gt){
    gs <- str_split(gt[,samp], "/|\\|", simplify = T) %>% 
      as.data.frame()
    
    pseudo_dip <- lapply(gs, function(x)(return(paste0(x, "|", x)))) %>% 
      bind_cols()
    
    colnames(pseudo_dip) <- paste0(samp, "_hap", 1:2)
    return(pseudo_dip)
  }
  
  ### GETTING THE PSEUDOHAPLOID MALE GT MATRIX ###
  pseudo_haploid_males_gts <- lapply(Males, hap_the_dip, gt = just_gt_Males) %>% 
    bind_cols()
  
  ### FUNCTION FOR GETTING PSEUDOHAPLOID FEMALES 
  hap_the_hap <- function(samp, gt){
    gs <- str_split(gt[,samp], "/|\\|", simplify = T) %>% 
      as.data.frame()
    
    pseudo_dip <- lapply(gs, function(x)(return(paste0(x, "|", x)))) %>% 
      bind_cols()
    pseudo_dip[pseudo_dip[,1] != pseudo_dip[,2],] <- ".|."
    pseudo_dip <- pseudo_dip[,1]
    colnames(pseudo_dip) <- paste0(samp, "_hapF")
    return(pseudo_dip)
  }
  
  ### GETTING THE PSEUDOHAPLOID FEMALE GT MATRIX ###
  pseudo_haploid_female_gts <- lapply(Females, hap_the_hap, gt = just_gt_Females) %>% 
    bind_cols()
  
  ### MERGING ###
  pseudo_haploid_gts <- cbind(data.frame(FORMAT = rep("GT:AD:DP"))) %>% 
                                cbind(pseudo_haploid_males_gts, pseudo_haploid_female_gts) %>% 
    as.matrix()
  pseudo_haploid_gts[pseudo_haploid_gts == "0|0"] <- "0|0:10,0:10"
  pseudo_haploid_gts[pseudo_haploid_gts == "1|1"] <- "1|1:0,10:10"
  pseudo_haploid_gts[pseudo_haploid_gts == ".|."] <- ".|.:0,0:0"
  
  pseudohaploid_vcf <- vcf_whole
  pseudohaploid_vcf@gt <- pseudo_haploid_gts
  
  newname <- gsub(".vcf.gz", "_pseudohaploid.vcf.gz", vcf)
  write.vcf(pseudohaploid_vcf, newname)
}  

```

```{bash engine.opts='-l'}
conda activate vcftools 
for vcf in data/vcfs/*pseudohaploid.vcf.gz; do
  gunzip $vcf
  bgzip -c ${vcf/.gz/} > $vcf
  tabix -p vcf $vcf
done
rm data/vcfs/*pseudohaploid.vcf
conda deactivate
```


```{r}
#Reading in genome file and running pooled stats
ChrZ_genomes <- c("NC_044241.2", "Z_filt.vcf.gz", "Z_nocoding_filt.vcf.gz", "Singhal_bTG1_4_Z_plusLTF_filtered_snps.vcf.gz")
names <- c("contig", "filt", "non_coding_filt", "gatk")
tps <- 30527588

for (i in 2:length(ChrZ_genomes)){
  genome <- readVCF(paste0("data/vcfs/", ChrZ_genomes[i]), numcols=10000, tid=ChrZ_genomes[1], frompos=1, topos=tps)
  genome <- sliding.window.transform(genome,width=100000,10000, type=2)
  
  ######## Running jakes stat version  ###############
  populations <- list(c(MAs[1:2], FAs), c(MBs, FBs))
  outgroup <- "LFT"
  genome <- set.populations(genome, populations, diploid = T)
  genome <- set.outgroup(genome,new.outgroup=outgroup,diploid = T)
  
  ## Calc stats ##
  genome <- F_ST.stats(genome, mode="nucleotide")
  pairwise.FST <- t(genome@nuc.F_ST.pairwise)
  positions <-seq(1, tps, (tps/nrow(pairwise.FST)))
                  
  all.stats_Chr <- bind_cols(positions, pairwise.FST) 
  colnames(all.stats_Chr) <- c("Pos", "FST")
  
  FST_plot <- all.stats_Chr %>% 
    ggplot(aes(x =Pos, y = FST)) + geom_line()
  ggsave(paste0("plots/jake/", names[i], "_FST_jake.pdf"), FST_plot)
  
  genome <- diversity.stats(genome, keep.site.info = F, pi = T)
  Pi <-  as.data.frame(genome@Pi)
  colnames(Pi) <- c("A_Pi", "B_Pi")
  Pi$n.sites <- genome@n.sites
  Pi$pos <- positions
  Pi$A_Pis <- Pi$A_Pi/Pi$n.sites
  Pi$B_Pis <- Pi$B_Pi/Pi$n.sites
  
  Pi_plot <- Pi %>% pivot_longer(cols = c(A_Pis, B_Pis), names_to = "pop", values_to = "Pi") %>% 
    ggplot(aes(x = pos, y =  Pi)) + geom_line() + 
    facet_wrap(~pop) 
  ggsave(paste0("plots/jake/", names[i], "_Pi_Jake.pdf"), Pi_plot)
}

```



```{r}
##### RUNNING MY PRETEND HAPLOID STATS ####
all <- c(FAs, MAs, paste0(MAs, ".2"), FBs, MBs, paste0(MBs, ".2"))
A_haps_ss <- c(FAs, MAs[1:2], paste0(MAs[1:2], ".2"))
B_haps <- c(FBs, MBs, paste0(MBs, ".2"))

A_haps_bad <- c(FAs, MAs[1:2], paste0(c(FAs, MAs[1:2]), ".2"))
B_haps_bad <- c(FBs, MBs, paste0(c(FBs, MBs), ".2"))

ChrZ_genomes <- c("NC_044241.2", "Z_filt.vcf.gz", "Z_nocoding_filt.vcf.gz", "Singhal_bTG1_4_Z_plusLTF_filtered_snps.vcf.gz")
names <- c("contig", "filt", "non_coding_filt", "gatk")
tps <- 30527588
populations1 <- list(A_haps_ss, B_haps)
populations2 <- list(A_haps_bad, B_haps_bad)
populations3 <- list(A_haps_ss, B_haps, A_haps_bad, B_haps_bad)
outgroup <- c("LFT", "LFT.2")

for (i in 2:length(ChrZ_genomes)){
  genome <- readVCF(paste0("data/vcfs/", ChrZ_genomes[i]), numcols=10000, tid=ChrZ_genomes[1], frompos=1, topos=tps)
  genome <- sliding.window.transform(genome,width=100000,10000, type=2)

  
  #Correct Chr number 
  genome <- set.populations(genome, populations1, diploid = F)
  genome <- set.outgroup(genome,new.outgroup=outgroup,diploid = F)
  
  ## Calc stats ##
  genome <- F_ST.stats(genome, mode="nucleotide")
  pairwise.FST <- t(genome@nuc.F_ST.pairwise)
  position <- data.frame(position = seq(1, tps, (tps/nrow(pairwise.FST))))
  all.stats_Chr <- bind_cols(position, pairwise.FST)
  colnames(all.stats_Chr) <- c("Pos", "FST")
  
  FST_plot_haploid <- all.stats_Chr %>% 
    ggplot(aes(x =Pos, y = FST)) + geom_line()
  ggsave(paste0("plots/pop_hap/", names[i], "_FST_pop_hap_corchrno.pdf"), FST_plot_haploid)
  
  
  #Incorrect Chr number
  genome <- set.populations(genome, populations2, diploid = F)
  genome <- set.outgroup(genome,new.outgroup=outgroup,diploid = F)
  
  ## Calc stats ##
  genome <- F_ST.stats(genome, mode="nucleotide")
  pairwise.FST <- t(genome@nuc.F_ST.pairwise)
  position <- data.frame(position = seq(1, tps, (tps/nrow(pairwise.FST))))
  all.stats_Chr <- bind_cols(position, pairwise.FST)
  colnames(all.stats_Chr) <- c("Pos", "FST")
  
  FST_plot_haploid <- all.stats_Chr %>% 
    ggplot(aes(x =Pos, y = FST)) + geom_line()
  ggsave(paste0("plots/pop_hap/", names[i], "_FST_pop_hap_incorchrno.pdf"), FST_plot_haploid)
  
  
  #All variations 
  genome <- set.populations(genome, populations3, diploid = F)
  genome <- set.outgroup(genome,new.outgroup=outgroup,diploid = F)
  
  genome <- diversity.stats(genome, keep.site.info = F, pi = T)
  Pi <-  as.data.frame(genome@Pi)
  colnames(Pi) <- c("Ahaps_pi", "Bhaps_pi", "Ahaps_bad_pi", "Bhaps_bad_pi")
  Pi$n.sites <- genome@n.sites
  Pi$pos <- positions
  Pi$Ahaps_pis <- Pi$Ahaps_pi/Pi$n.sites
  Pi$Bhaps_pis <- Pi$Bhaps_pi/Pi$n.sites
  Pi$Ahaps_bad_pis <- Pi$Ahaps_bad_pi/Pi$n.sites
  Pi$Bhaps_bad_pis <- Pi$Bhaps_bad_pi/Pi$n.sites
  
  Pi_plot <- Pi %>% pivot_longer(cols = c(Ahaps_pis, Bhaps_pis, Ahaps_bad_pis, Bhaps_bad_pis), names_to = "pop", values_to = "Pi") %>% 
   ggplot(aes(x = pos, y =  Pi)) + geom_line() + 
    facet_wrap(~pop) 
  ggsave(paste0("plots/pop_hap/", names[i], "_Pi_pop_hap.pdf"), Pi_plot)
}
```



```{r}
#Reading in genome file and running pooled stats

ChrZ_genomes <- c("NC_044241.2", "Z_filt_pseudohaploid.vcf.gz", "Z_nocoding_filt_pseudohaploid.vcf.gz",
                  "Singhal_bTG1_4_Z_plusLTF_filtered_snps_pseudohaploid.vcf.gz")
names <- c("contig", "filt", "non_coding_filt", "gatk")
tps <- 30527588

for (i in 2:length(ChrZ_genomes)){
  genome_ph <- readVCF(paste0("data/vcfs/", ChrZ_genomes[i]), numcols=10000, tid=ChrZ_genomes[1], frompos=1, topos=tps)
  genome_ph <- sliding.window.transform(genome_ph,width=100000,10000, type=2)

  Ahaps <- grep(paste(c(FAs, MAs[1:2]), collapse="|"), colnames(pseudo_haploid_gts), value=T)
  Ahaps_bad <- c(Ahaps, Ahaps[c(5:8)])
  
  Bhaps <- grep(paste(c(FBs, MBs), collapse="|"), colnames(pseudo_haploid_gts), value=T)
  Bhaps_bad <- c(Bhaps, Bhaps[c(3,4,5,6,7,8)]) 
  
  population1 <- list(Ahaps, Bhaps)
  population2 <- list(Ahaps_bad, Bhaps_bad)
  population3 <- list(Ahaps, Bhaps, Ahaps_bad, Bhaps_bad)
  
  outgroup <- c("LFT_hap1", "LFT_hap2")
  
  ###### Correct Chr number  ###### 
  genome_ph <- set.populations(genome_ph, population1, diploid = F)
  genome_ph <- set.outgroup(genome_ph,new.outgroup=outgroup,diploid = F)
  
  ## Calc stats ##
  genome_ph <- F_ST.stats(genome_ph, mode="nucleotide")
  pairwise.FST <- t(genome_ph@nuc.F_ST.pairwise)
  positions <- data.frame(position = seq(1, tps, (tps/nrow(pairwise.FST))))
  all.stats_Chr <- bind_cols(position, pairwise.FST)
  colnames(all.stats_Chr) <- c("Pos", "FST")
  
  FST_plot_haploid <- all.stats_Chr %>% 
    ggplot(aes(x =Pos, y = FST)) + geom_line()
  ggsave(paste0("plots/pseudo//", names[i], "_FST_pseudo_corchrno.pdf"), FST_plot_haploid)
  
  
  ###### Incorrect Chr number ####
  genome_ph <- set.populations(genome_ph, population2, diploid = F)
  genome_ph <- set.outgroup(genome_ph,new.outgroup=outgroup,diploid = F)
  
 ## Calc stats ##
  genome_ph <- F_ST.stats(genome_ph, mode="nucleotide")
  pairwise.FST <- t(genome_ph@nuc.F_ST.pairwise)
  positions <- seq(1, tps, (tps/nrow(pairwise.FST)))
  all.stats_Chr <- bind_cols(position, pairwise.FST)
  colnames(all.stats_Chr) <- c("Pos", "FST")
  
  FST_plot_haploid <- all.stats_Chr %>% 
    ggplot(aes(x =Pos, y = FST)) + geom_line()
  ggsave(paste0("plots/pseudo//", names[i], "_FST_pseudo_incorchrno.pdf"), FST_plot_haploid)
  
  
  #########################
  
  genome_ph <- set.populations(genome_ph, population3, diploid = F)
  genome_ph <- set.outgroup(genome_ph,new.outgroup=outgroup,diploid = F)
  
  genome_ph <- diversity.stats(genome_ph, pi=T, keep.site.info = F)
  Pi <-  as.data.frame(genome_ph@Pi)
  colnames(Pi) <- c("Ahaps_pi", "Bhaps_pi", "Ahaps_bad_pi", "Bhaps_bad_pi")

  Pi$n.sites <- genome_ph@n.sites
  Pi$pos <- positions
  Pi$Ahaps_pis <- Pi$Ahaps_pi/Pi$n.sites
  Pi$Bhaps_pis <- Pi$Bhaps_pi/Pi$n.sites
  Pi$Ahaps_bad_pis <- Pi$Ahaps_bad_pi/Pi$n.sites
  Pi$Bhaps_bad_pis <- Pi$Bhaps_bad_pi/Pi$n.sites
  
  Pi_plot <- Pi %>% pivot_longer(cols = c(Ahaps_pis, Bhaps_pis, 
                                          Ahaps_bad_pis, Bhaps_bad_pis), 
                                 names_to = "pop", values_to = "Pi") %>% 
   ggplot(aes(x = pos, y =  Pi)) + geom_line() + 
    facet_wrap(~pop) 
  ggsave(paste0("plots/pseudo/", names[i], "_Pi_pseudo.pdf"), Pi_plot)

}
```

